### PR TITLE
rename context to serialization_context and add url helpers

### PR DESCRIPTION
**Pull request by tchak**

### CODE REVIEW

**Comment by tchak** [comment]
I am not sure about the best way to include url_helpers in the context object. If some people with more knowledge about rails internals have a better idea I am all :ear:  :)

**Comment by tchak** [comment]
I also feel that in the end `serialization_context` should be exposed just like `scope` directly on the serializer instances. I would like to add `serialization_context_name` just like for `scope` and expose it as `context` by default.

File: lib/action_controller/serialization/context.rb [hunk 1]
```diff
@@ -0,0 +1,18 @@
+module ActionController
+  module Serialization
+    class Context
+      include Rails.application.routes.url_helpers if Rails.application.present?
+
+      def self.default_url_options
+        ActionMailer::Base.default_url_options
+      end
```
Comment (bf4) [inline lib/action_controller/serialization/context.rb]: I'd rather not include the url helpers and instead stick 'em in a url_helpers method

We could probably set it and default_url_iptions in the railtie.

Also, i don't believe we include action mailer nor do we need it. There should be a similar methid on the controller

**Comment by bf4** [comment]
Awesome work!

File: lib/action_controller/serialization/context.rb (hunk 1 above)
Comment (bf4) [inline lib/action_controller/serialization/context.rb]: Let's also add a context test

Comment (tchak) [inline lib/action_controller/serialization/context.rb]: Oh sorry about `ActionMailer` this is me trying things by copy/pasting from SO in a hurry :( Will look in to how do it from the railtie!

**Comment by beauby** [comment]
Pretty cool :+1:

**Comment by NullVoxPopuli** [comment]
LGTM :+1:

**Comment by NullVoxPopuli** [comment]
@tchak I'm good with merging this, unless you think `serialization_context_name` should be included in this PR.. maybe a different PR? All tests pass, so... :-)

**Comment by tchak** [comment]
@NullVoxPopuli let me add a few more test tonight. Yes `serialization_context_name` will go in a separate PR.

**Comment by NullVoxPopuli** [comment]
cool. thanks!

File: lib/action_controller/serialization/context.rb [hunk 2]
```diff
@@ -0,0 +1,13 @@
+module ActionController
+  module Serialization
+    class Context
+      attr_reader :request_url, :query_parameters, :url_helpers
```
Comment (bf4) [inline lib/action_controller/serialization/context.rb]: I think I'd rather have `url_helpers` a class method on the Context.  It feels more natural here. Also, no need to include them. you could just

```ruby
class Context
  class << self
    attr_accessor :url_helpers, :default_url_options
  end
end
```

and in the Railtie

```ruby
Context.url_helpers = Rails.application.routes.url_helpers # for now, later would like to encapsulate it to control the boundary e.g. UrlHelpers.new(Rails.application.routes.url_helpers)
Context.default_url_options =  ActionController::Base.default_url_options
```

I'm trying to keep in mind how someone using grape would configure this.  If we keep a good interface, the details of the implementation don't matter as much.  That means more composition and less just mixing things in. Alternatively, there could be a UrlHelpers module that you mix in to the Context when the railtie is called, which allows others to mix in a different helper object

File: lib/active_model_serializers.rb [hunk 3]
```diff
@@ -7,6 +7,8 @@ module ActiveModelSerializers
   mattr_accessor :logger
   self.logger = Rails.logger || Logger.new(IO::NULL)
 
+  mattr_accessor :url_helpers
```
Comment (bf4) [inline lib/active_model_serializers.rb]: see https://github.com/rails-api/active_model_serializers/pull/1289/files#r42808317

File: lib/active_model_serializers.rb [hunk 4]
```diff
@@ -7,6 +7,8 @@ module ActiveModelSerializers
   mattr_accessor :logger
   self.logger = Rails.logger || Logger.new(IO::NULL)
```
Comment (bf4) [inline lib/active_model_serializers.rb]: This should really be set to Rails.logger in the Railtie, not here. womp womp

File: lib/action_controller/serialization/context.rb (hunk 2 above)
Comment (tchak) [inline lib/action_controller/serialization/context.rb]: hmm as far as I understand `default_url_options` needs to be defined on the same object where `url_helpers` are included in order to be taken in to account. Am I not understanding it correctly?

Comment (tchak) [inline lib/action_controller/serialization/context.rb]: also I usually use url helpers in serialisers a lot, so I am thinking that at some point I would like to be able to do in some serializer code something like this:
```
def validation_url
  context.validation_url(object)
end
```

Are we on the same page here? Or do you see the usage differently?

**Comment by tchak** [comment]
I am trying to add some tests to `url_helpers` thing and I feel a bit confused :) I am going to need a better understanding of how rails routes works and compose with everything else... I hope to be able to figure it out in the next few days.

**Comment by beauby** [comment]
In my opinion this PR is worth merging without url_helpers (that we could add later).

**Comment by NullVoxPopuli** [comment]
I agree, @beauby

**Comment by beauby** [comment]
Wait, it wasn't ready though (there were references to url_helpers and stuff). Could you revert the merge until we clear it for merging?

**Comment by NullVoxPopuli** [comment]
that's what I get for not re-reviewing. ha

**Comment by bf4** [comment]
@tchak would you mind re-submitting this PR, referencing it, and referring to the comments in https://github.com/rails-api/active_model_serializers/pull/1289#discussion_r42808317 ? Maybe skip the do it in two parts, without the url helpers, then adding them

**Comment by beauby** [comment]
So in the end, how does one access the `url_helpers` from within an attribute definition in a serializer?
As follows?
```ruby
attribute :url do
  serialization_context.url_helpers.posts_route
end
```

**Comment by bf4** [comment]
I think I left this open for the uri templates

The serialization context url helpers are mixed into Link.  Examples in docs

B mobile phone

**Pull request merged**